File: examples/python/lib/rpc.py
import json, requests, uuid
from .hash   import hash160
from .helper import decode_address

class RpcSocket:
    ''' Basic implementation of a JSON-RPC interface. '''
    def __init__(self, opt):
        url  = opt.get('url', '127.0.0.1')
        port = opt.get('port', 18443)

        self.fullUrl  = f'http://{url}:{port}/'
        self.username = opt.get('username', 'bitcoin')
        self.password = opt.get('password', 'password')
        self.wallet   = opt.get('wallet', None)
        self.initFlag = False

    def init(self):
        ''' Initialize the RPC object. '''
        if self.wallet:
            self.loadWallet()
            self.fullUrl += f'wallet/{self.wallet}'
        

    def call(self, method, args = []):
        ''' Make an RPC call using the 
            specified method and arguments.
        '''
        # Make sure to initialize the RPC object first.
        if not self.initFlag:
            self.initFlag = True
            self.init()
        
        # Format the arguments before the call.
        args = args if type(args) is list else [ args ]

        # Construct the body of the reuqest.
        body = json.dumps({
            "jsonrpc": "1.0",
            "id": uuid.uuid4().urn.split(':')[-1],
            "method": method,
            "params": args
        })
        
        # Make the request to the server.
        response = requests.post(
            self.fullUrl,
            auth=(self.username, self.password),
            data=body
        )
        
        # If the response code is not 200, fail here.
        if response.status_code != 200:
            raise Exception(f'Response failed with error: {response.status_code}')
        
        data = response.json()
        
        # If the data includes an error message, fail here.
        if data['error']:
            if data['error']['code'] == -1:
                raise Exception(f'RPC command {method} failed with syntax error. Please check your arguments.')
            else:
                err_msg = data["error"]["message"]
                raise Exception(f'RPC command {method} failed with error: {err_msg}')
        
        return data['result']

    def check(self):
        res = self.call('getblockchaininfo')
        return 'chain' in res


    def isWalletLoaded(self):
        wallets = self.call('listwallets')
        return self.wallet in wallets


    def isWalletExists(self):
        wallets = self.call('listwalletdir')
        return [ w['name'] for w in wallets if w['name'] == self.wallet ]


    def loadWallet(self):
        if self.isWalletLoaded():
            return True
        if not self.isWalletExists():
            raise f'Wallet "{self.wallet}" does not exist!'

        result = self.call('loadwallet', [ self.wallet ])

        if result['warning'] or result['name'] != self.wallet:
            raise 'Wallet failed to load cleanly: {}'.format(result['warning'])
        
        return True

    def get_utxo(self, idx):
        utxos = self.call('listunspent')
        if len(utxos) <= idx:
          raise Exception('You are requesting an index out of range!')
        address = utxos[idx]['address']
        encoded_key = self.call('dumpprivkey', address)
        public_key  = self.call('getaddressinfo', address)['pubkey']
        pubkey_hash = hash160(public_key).hex()

        return {
            'address': address,
            'priv_key': decode_address(encoded_key),
            'pub_key': public_key,
            'pubkey_hash': pubkey_hash,
            'redeem_script': f'1976a914{pubkey_hash}88ac',
            'txid': utxos[idx]['txid'],
            'vout': utxos[idx]['vout'],
            'value': int(utxos[idx]['amount'] * 100000000)
        }

    def get_recv(self, address=None):
        if address is None:
            address = self.call('getnewaddress')
        encoded_key = self.call('dumpprivkey', address)
        public_key  = self.call('getaddressinfo', address)['pubkey']
        pubkey_hash = hash160(public_key).hex()
        return {
            'address': address,
            'priv_key': decode_address(encoded_key),
            'pub_key': public_key,
            'pubkey_hash': pubkey_hash,
            'redeem_script': f'1976a914{pubkey_hash}88ac',
        }

File: examples/node/lib/rpc.mjs
#!/usr/bin/env -S node --experimental-modules --no-warnings --trace-uncaught

import { webcrypto as crypto } from 'crypto'
import { fileURLToPath } from 'url';

// Unpack arguments from the current process.
const [ nodepath, sourcepath, method, ...params ] = process.argv

// Check if this script is being executed from shell.
const filepath = fileURLToPath(import.meta.url)
const isExecutedFromShell = Boolean(sourcepath === filepath)

if (isExecutedFromShell) {
  if (!method) {
    // If method is blank, show help text and exit.
    const filename = sourcepath.split('/')[-1]
    console.log(
      `Usage: ./${filename} [ METHOD ] [ PARAMS ]\n\n` +
      'Basic implementation of using json-rpc with node.js\n\n' +
      `Example: ./${filename} getblockchaininfo`
    )
    process.exit(0)
  } else {
    // Else, print our results.
    const result = await rpc(method, params)
    console.log('Result:', result)
  }
}

export default async function rpc(method, args = [], config = {}) {
  /** Send a JSON-RPC call to the configured server. */

  const user = config.user   || 'bitcoin',    // RPC-Auth Username
        pass = config.pass   || 'password',   // RPC-Auth Password
        url  = config.url    || '127.0.0.1',  // URL to your Bitcoin node.
        port = config.port   || 18443         // Port to your RPC interface.

  let wallet = config.wallet || ''            // Name of wallet to use.

  // Random identifer for our request.
  const requestId = crypto.randomUUID()

  // Authorization string for our request.
  const authString = Buffer.from(user + ':' + pass).toString('base64')

  // Make sure our args are in an array.
  args = (Array.isArray(args)) ? args : [ args ]

  try {
  
    if (wallet) {
      // If a wallet is specified, ensure that the wallet file
      // exists and is loaded, then configure the url endpoint.
      await loadWallet(wallet)
      wallet = 'wallet/' + wallet
    }

    // Confgigure our request object.
    const request = {
      method: 'POST',
      headers: {
        'Authorization': 'Basic ' + authString,
        'content-type': 'application/json'
      },
      body: JSON.stringify({
        "jsonrpc": "1.0",
        "id": requestId,
        "method": method,
        "params": args
      })
    }

    // Fetch a response from our node.
    const response = await fetch(`http://${url}:${port}/${wallet}`, request)

    // If the response fails, throw an error.
    if (!response.ok && !response.json) {
      throw `Request for '${method}' failed with status ${response.status}: ${response.statusText}`
    }

    // Convert our response to json.
    const { result, error } = await response.json()

    // If the RPC call has an error, unpack and throw the error.
    if (error) {
      const { code, message } = error
      if (code === -1) {
        throw `RPC command ${method} failed with syntax error. Please check your arguments.`
      } else { throw `RPC command ${method} failed with error: ${message}` }
    }

    return result

  } catch(err) { throw err }
}

async function isWalletLoaded(walletName) {
  /** Check if the specified wallet is loaded
   *  within the bitcoin-core node.
   * */
  return rpc('listwallets')
    .then((wallets) => wallets && wallets.includes(walletName))
}

async function isWalletExists(walletName) {
  /** Check if the specified wallet exists within 
   *  the host filesystem for bitcoin-core.
   * */
  return rpc('listwalletdir')
    .then(({ wallets }) => wallets.find(el => el.name === walletName))
}

async function loadWallet(walletName) {
  /** Ensure that the specified wallet is loaded for 
   *  the bitcoin-core node and available to access.
   * */

  if (await isWalletLoaded(walletName)) {
    // If wallet is already loaded, return.
    return true
  }

  if (!(await isWalletExists(walletName))) {
    // If wallet does not exist, throw error.
    throw 'Wallet file does not exist!'
  }

  return rpc('loadwallet', [ walletName ])
    .then(({ name, warning }) => {
      if (warning || name !== walletName) {
        // If there was a problem with loading, throw error.
        throw `Wallet failed to load cleanly: ${warning}`
      }
      return true
    })
}
